### PR TITLE
Adds qn_targets_available Endpoint

### PR DESCRIPTION
**Pull request by Miserlou**
## Issue Number
https://github.com/AlexsLemonade/refinebio/issues/644

One possible solution to the 644 problem, allows us to search/return all samples but gives frontend a method to filter/hide un-QN'able samples.

### CODE REVIEW

**Comment by arielsvn** [comment]
LGTM. I'm gonna try to implement the frontend changes before we merge this to ensure we cover all cases.

**Comment by arielsvn** [comment]
Sample response: 

```
[
  {'name': 'HOMO_SAPIENS', 'taxonomy_id': 9606}, 
  {'name': 'DANIO_RERIO', 'taxonomy_id': 7955}
]
```

**Comment by arielsvn** [comment]
@Miserlou added https://github.com/AlexsLemonade/refinebio-frontend/pull/559 that ensures samples without QN targets are displayed as not processed in the samples table.

There're two details that I think we should add to the API:

1. Ensure that only samples with QN targets can be added to the dataset. This is especially important when we add entire experiments with the key `ALL`

2. Send the number of downloadable samples: samples that are processed and have QN targets. We need to display this number both in the search page and the experiments page.

image

As we discussed, we need to think what would be the best way to add this.

**Comment by Miserlou** [comment]
I strongly agree with (1), although the result object should be the same at smashtime, they will get the "unsmashable files" warning for items they didn't add, I think (2) might be harder and that might not be the correct way to solve it, my gut says that should be turned into something cachable.

**Comment by Miserlou** [comment]
The ['ALL'] issue is addressed now, (2) isn't but might be out of scope for this PR as it requires some additional architecture/API design consideration.

File: api/data_refinery_api/views.py [hunk 1]
```diff
@@ -540,7 +541,12 @@ def perform_update(self, serializer):
             accessions = new_data['data'][key]
             if accessions == ["ALL"]:
                 experiment = get_object_or_404(Experiment, accession_code=key)
-                sample_codes = list(experiment.samples.filter(is_processed=True).values_list('accession_code', flat=True))
+
+                # QN organisms
+                organism_ids = list(ComputationalResultAnnotation.objects.filter(data__is_qn=True).values_list('data__organism_id', flat=True).order_by('data__organism_id'))
+                organisms = Organism.objects.filter(id__in=organism_ids)
```
Comment (arielsvn) [inline api/data_refinery_api/views.py]: What about calculating this outside of the `for...loop` above? In some cases we send multiple `ALL` keys:

image

File: api/data_refinery_api/views.py [hunk 2]
```diff
@@ -540,7 +541,12 @@ def perform_update(self, serializer):
...
+
+                sample_codes = list(experiment.samples.filter(is_processed=True, organism__in=organisms).values_list('accession_code', flat=True))
```
Comment (arielsvn) [inline api/data_refinery_api/views.py]: Should we do the same check for when someone wants to add a sample with an accession code? And throw an error if the sample doesn't have any QN targets associated?

File: api/data_refinery_api/views.py [hunk 3]
```diff
@@ -1336,6 +1342,19 @@ def get(self, request, format=None):
...
 
+class QNTargetsAvailable(APIView):
+    """
+    This is a list of all of the organisms which have available QN Targets
+    """
+    def get(self, request, format=None):
+        
+
+        organism_ids = list(ComputationalResultAnnotation.objects.filter(data__is_qn=True).values_list('data__organism_id', flat=True).order_by('data__organism_id'))
+        organisms = Organism.objects.filter(id__in=organism_ids)
```
Comment (arielsvn) [inline api/data_refinery_api/views.py]: Seems that we use this in multiple places, is it worthwhile to move it into `Organism`? Similarly to `Experiments.public_objects`

File: api/data_refinery_api/views.py (hunk 2 above)
Comment (Miserlou) [reply api/data_refinery_api/views.py]: That's an interesting question but I'm not sure it's the API's duty to do that at that stage if we're returning them those objects only to have using them be a source of failure..

**Comment by Miserlou** [comment]
I'd like to get this reviewed and merged as it will fix some of the sporadic failures in other PRs.

File: api/data_refinery_api/views.py [hunk 4]
```diff
@@ -1357,6 +1361,19 @@ def get(self, request, format=None):
...
```
Comment (arielsvn) [inline api/data_refinery_api/views.py]: `Organism.get_objects_with_qn_targets()` ?

**Pull request merged**